### PR TITLE
gzipped FASTA input support

### PR DESCRIPTION
**Pull request by nekrut**

### CODE REVIEW

**Comment by peterjc** [comment]
Cross reference #98, I will merge that now.

**Comment by peterjc** [comment]
This includes support for gzip'ed FASTA input - see #101

**Comment by peterjc** [comment]
Judging from the cron jobs, the failing TravisCI tests on the master are in part upstream changes and/or other issues with the dependencies stack. Sigh.

**Comment by nekrut** [comment]
@peterjc please take a look at makeblastdb `<command>` section to see if my changes are reasonable.

**Comment by peterjc** [comment]
@nekrut I'm hoping to pull in your changes by hand (keeping authorship intact), and ought to be able to test locally even if TravisCI is giving me trouble. Thanks for working on this.

See also https://github.com/galaxyproject/galaxy/issues/6180 which is one of the issues at play.

**Comment by nekrut** [comment]
@peterjc I don't care about the authorship as long as things work, so do whatever is easiest

**Comment by peterjc** [comment]
Maintaining git authorship is simple enough, I just need to get my master branch tests working again...

https://github.com/peterjc/galaxy_blast/pull/103 seems to have fixed testing with the Galaxy dev branch :)

**Comment by peterjc** [comment]
Rebased, squashed a few commits, removed the change to the version scheme, and updated the README.

Let's see if TravisCI is happy, then think about adding tests for building databases from gzipped FASTA files. I note that as written this adds additional dependencies on the magic library...

**Comment by peterjc** [comment]
The TIGRFAM URL seems to be causing more trouble, see also https://twitter.com/pjacock/status/998941086168076288 and https://github.com/peterjc/galaxy_blast/commit/d31e21d684a5dfa55dfb1a9aead744e2e7b87923

**Comment by peterjc** [comment]
Still a couple of tests failing,

```
Exception: Request to upload dataset failed [{"err_data": {"file_type": "An invalid option was selected for file_type, u'fasta.gz', please verify."}, "traceback": "Traceback (most recent call last):\n  File \"/home/travis/build/peterjc/galaxy_blast/galaxy-master/lib/galaxy/web/framework/decorators.py\", line 281, in decorator\n    rval = func(self, trans, *args, **kwargs)\n  File \"/home/travis/build/peterjc/galaxy_blast/galaxy-master/lib/galaxy/webapps/galaxy/api/tools.py\", line 415, in create\n    return self._create(trans, payload, **kwd)\n  File \"/home/travis/build/peterjc/galaxy_blast/galaxy-master/lib/galaxy/webapps/galaxy/api/tools.py\", line 473, in _create\n    vars = tool.handle_input(trans, incoming, history=target_history, use_cached_job=use_cached_job)\n  File \"/home/travis/build/peterjc/galaxy_blast/galaxy-master/lib/galaxy/tools/__init__.py\", line 1370, in handle_input\n    raise exceptions.MessageException(', '.join(msg for msg in err_data.values()), err_data=err_data)\nMessageException: An invalid option was selected for file_type, u'fasta.gz', please verify.\n", "err_msg": "An invalid option was selected for file_type, u'fasta.gz', please verify.", "err_code": 0}]
```

We need to update ``.travis.datatypes_conf.xml`` (a quirk of my TravisCI setup from testing datatypes which were outside the Galaxy core), I'll do that next...

**Comment by peterjc** [comment]
That took a while - does this look OK now to merge from your point of view @nekrut? Thanks!

**Comment by peterjc** [comment]
If I don't hear otherwise this week, I aim to merge this.

The main gotcha (which the test suite caught) was reverting some of the quoting of arguments in the wrapper (normally good practise in Galaxy wrappers, but here can't be used in a couple of cases where a UI option is mapped to multiple command line switches).

File: tools/ncbi_blast_plus/ncbi_blastn_wrapper.xml [hunk 1]
```diff
@@ -7,24 +7,29 @@
...
 @THREADS@
 #if $adv_opts.adv_opts_selector=="advanced":
-$adv_opts.strand
+${adv_opts.strand}
 @ADV_FILTER_QUERY@
 @ADV_MAX_HITS@
 @ADV_WORD_SIZE@
 #if (str($adv_opts.identity_cutoff) and float(str($adv_opts.identity_cutoff)) > 0 ):
--perc_identity $adv_opts.identity_cutoff
+-perc_identity '${adv_opts.identity_cutoff}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_blastn_wrapper.xml]: Don't need quotes here, `identity_cutoff` is a `float`.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 2]
```diff
@@ -533,42 +533,46 @@
         </conditional>
     </xml>
 <!--Tokens-->
-    <token name="@ADV_MATRIX_GAPCOSTS@">
+    <token name="@ADV_MATRIX_GAPCOSTS@"><![CDATA[
 #if str($adv_opts.matrix_gapcosts.matrix):
-    -matrix $adv_opts.matrix_gapcosts.matrix
-    $adv_opts.matrix_gapcosts.gap_costs
+    -matrix '${adv_opts.matrix_gapcosts.matrix}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 3]
```diff
@@ -533,42 +533,46 @@
...
 #end if
-    </token>
+    ]]></token>
 
-    <token name="@ADV_QCOV_HSP_PERC@">
-#if float(str($adv_opts.qcov_hsp_perc)) &gt; 0:
-    -qcov_hsp_perc $adv_opts.qcov_hsp_perc
+    <token name="@ADV_QCOV_HSP_PERC@"><![CDATA[
+#if float(str($adv_opts.qcov_hsp_perc)) > 0:
+    -qcov_hsp_perc '${adv_opts.qcov_hsp_perc}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 4]
```diff
@@ -579,79 +583,81 @@
...
-    <token name="@ADV_MAX_HITS@">
+    <token name="@ADV_MAX_HITS@"><![CDATA[
 ## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
 ## Note -max_target_seqs used to simply override -num_descriptions and -num_alignments
 ## but this was changed in BLAST+ 2.2.27 onwards to force their use (raised with NCBI)
 #if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
     #if str($output.out_format) in ["6", "ext", "cols", "5"]:
         ## Most output formats use this, including tabular and XML:
-        -max_target_seqs $adv_opts.max_hits
+        -max_target_seqs '${adv_opts.max_hits}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 5]
```diff
@@ -579,79 +583,81 @@
...
     #else
         ## Text and HTML output formats 0-4 currently need this instead:
         -num_descriptions $adv_opts.max_hits -num_alignments $adv_opts.max_hits
     #end if
 #end if
 #if str($adv_opts.max_hsps)
-    -max_hsps $adv_opts.max_hsps
+    -max_hsps '${adv_opts.max_hsps}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 6]
```diff
@@ -579,79 +583,81 @@
...
 #end if
-    </token>
-    <token name="@ADV_WORD_SIZE@">
+    ]]></token>
+    <token name="@ADV_WORD_SIZE@"><![CDATA[
 #if str($adv_opts.word_size):
--word_size $adv_opts.word_size
+-word_size '${adv_opts.word_size}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 7]
```diff
@@ -579,79 +583,81 @@
...
 #end if
 $adv_opts.parse_deflines
-    </token>
-    <token name="@ADV_WINDOW_SIZE@">
+    ]]></token>
+    <token name="@ADV_WINDOW_SIZE@"><![CDATA[
 #if str($adv_opts.window_size):
--window_size $adv_opts.window_size
+-window_size '${adv_opts.window_size}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 8]
```diff
@@ -579,79 +583,81 @@
...
 #end if
-    </token>
+    ]]></token>
 
-    <token name="@ADV_THRESHOLD@">
+    <token name="@ADV_THRESHOLD@"><![CDATA[
 #if str($adv_opts.threshold):
--threshold $adv_opts.threshold
+-threshold '${adv_opts.threshold}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 9]
```diff
@@ -579,79 +583,81 @@
...
 #end if
-    </token>
+    ]]></token>
 
-    <token name="@ADV_COMP_BASED_STATS@">
+    <token name="@ADV_COMP_BASED_STATS@"><![CDATA[
 #if str($adv_opts.comp_based_stats):
--comp_based_stats $adv_opts.comp_based_stats
+-comp_based_stats '${adv_opts.comp_based_stats}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 10]
```diff
@@ -579,79 +583,81 @@
...
 #end if
-    </token>
+    ]]></token>
 
-    <token name="@ADV_GAPOPEN@">
+    <token name="@ADV_GAPOPEN@"><![CDATA[
 #if str($adv_opts.gapopen):
--gapopen $adv_opts.gapopen
+-gapopen '${adv_opts.gapopen}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 11]
```diff
@@ -579,79 +583,81 @@
...
 #end if
-    </token>
+    ]]></token>
 
-    <token name="@ADV_GAPEXTEND@">
+    <token name="@ADV_GAPEXTEND@"><![CDATA[
 #if str($adv_opts.gapextend):
--gapextend $adv_opts.gapextend
+-gapextend '${adv_opts.gapextend}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 12]
```diff
@@ -579,79 +583,81 @@
...
 #end if
-    </token>
+    ]]></token>
 
-    <token name="@ADV_MATRIX@">
+    <token name="@ADV_MATRIX@"><![CDATA[
 #if str($adv_opts.matrix):
--matrix $adv_opts.matrix
+-matrix '${adv_opts.matrix}'
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Quotes not needed.

File: tools/ncbi_blast_plus/ncbi_makeblastdb.xml [hunk 13]
```diff
@@ -46,8 +55,8 @@ $hash_index
 #end if
 ## --------------------------------------------------------------------
 ## Capture the stdout log information to the primary file (plain text):
-&gt; "$outfile"
-    </command>
+> "$outfile"
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_makeblastdb.xml]: Use single quotes.

Comment (peterjc) [reply tools/ncbi_blast_plus/ncbi_makeblastdb.xml]: Oh, good catch on the outfile quotes.

**Comment by peterjc** [comment]
@nsoranzo You flagged a lot of the quoting Anton added as unnecessary - does the IUC have any thing more explicit to say on this beyond:

> All Cheetah variables for text parameters, input and output files must be single-quoted, e.g. '${var_name}'.

https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html

**Comment by nsoranzo** [comment]
Not really, it's probably more of a *de facto* standard. I guess using quotes only when needed makes the final command shorter/cleaner and may help to spot errors.

**Comment by peterjc** [comment]
Thanks @nekrut - merged during the GCCBOSC 2018 CollaborationFest

**Pull request merged**